README.md
# DNS Monitor on Single Host

### Functionality

Summary: monitor outbound DNS traffic on a single host and process it in a streaming manner

1. Monitor outbound DNS queries through `tcpdump`
2. Deduplicate records such as adjacent `A` and `AAAA` records with the same domain name
3. Filter out special record types such as `PTR`
4. Aggregate information from all previously monitored queries
    - How many times the top-level domain (TLD) in the current record is queried
    - How many times the second-level domain (SLD) in the current record is queried
    - Select SLD or TLD as the domain key to print together with the full domain name
    - Color the domain key before printing to stdout regarding how many times it is queried
5. Handle all types of errors and print them to stderr asynchronously
    - Errors from `tcpdump` (the source)
    - Errors in information extraction (invalid records)

### Compilation Environment

[Haskell Stack](https://www.haskellstack.org/)

The latest stack is recommended: if installed with `apt-get` or `dnf`, then stack can be upgraded with `stack upgrade`. The latest stack is usually at `~/.local/bin/stack`.

### Installation
The application requires `tcpdump` installed on the host.
```
make install
```
This command compiles it, copies it to `/usr/bin` and sets appropriate permissions for it to run.

### Usage
```
dnsmonitor
```

### Screenshot

<img src="screenshot/1.0.png" width="760">

Color coding is based on TLD/SLD hit count ([Config.hs](src/Config.hs)):
```Haskell
colorHit :: Integer -> Text
colorHit h
    | h == 1    = "yellow"
    | h < 16    = "default"
    | h < 128   = "grey"
    | otherwise = "black"
```


test/ChangeLog.md
# Changelog for dnsmonitor-host

## Unreleased changes


src/Monitor.hs
{-# LANGUAGE OverloadedStrings #-}

module Monitor
( monitor
) where

import Config

import qualified Streamly.Prelude as S
import qualified Data.Text as T
import qualified Data.Text.IO as TIO
import qualified Data.HashMap.Strict as HashMap

import Streamly (runStream, asyncly)
import Control.Monad (guard, when)
import Data.Maybe (isJust, isNothing, fromJust)
import System.IO (Handle, stderr)
import System.Process (createProcess, proc, std_out, std_err, StdStream(..))

data TRD = TRD
    { timestamp  :: T.Text  
    , recordtype :: T.Text  
    , domainname :: T.Text  
    }

data RecTypeFilterState = RTFS
    { prevDomain :: T.Text
    , currTRD :: TRD
    }

data DomainProcState = DPS
    { tldMap :: DomainHitMap -- top-level domain
    , sldMap :: DomainHitMap -- second-level domain
    , currentTLD :: Maybe T.Text -- Just TLD
    , currentSLD :: Maybe T.Text -- Just SLD
    , currentdomain :: T.Text -- entire domain
    }

type DomainHitMap = HashMap.HashMap T.Text Integer

emptyDHM = DPS HashMap.empty HashMap.empty Nothing Nothing ""
emptyRTFS = RTFS "" $ TRD "" "" ""

monitor :: IO ()
monitor = do
    -- DNS data Producer from tcpdump
    (_, Just hout, Just herr, _) <- createProcess
        (proc "tcpdump" ["-i", "any", "-l", "-nn", "dst port 53"])
        { std_out = CreatePipe
        , std_err = CreatePipe }
    -- streaming DNS data, process it, and print it
    runStream . asyncly
        $  (printS . analyzeS . dupF . rtF . genTRDs $ S.fromHandle hout)
        <> (S.mapM (printErr . T.pack) $ S.fromHandle herr)
  where
    -- generate TRD from raw tcpdump records
    genTRDs    = filterJust . S.mapM generateTRD
    -- filter out TRDs based on record types
    rtF        = S.filter $ monitoredRecType . domainname
    -- filter out duplicate TRDs which share same domain names
    dupF       = filterJust . S.scanx updateRTFS emptyRTFS eliminateTRD
    -- stateful analysis to output line for print
    analyzeS   = filterJust . S.scanx updateHitMap emptyDHM outputLine
    -- display line
    printS     = S.mapM TIO.putStrLn
    -- helper function with type (Maybe a) -> a
    filterJust = fmap fromJust . S.filter isJust

generateTRD :: String -> IO (Maybe TRD)
generateTRD rawline = do
    when (isNothing mtrd) $ printErrInvalidRec line
    return mtrd
  where
    line = T.pack rawline
    mtrd = extractTRD line

extractTRD :: T.Text -> Maybe TRD
extractTRD line = do
    guard (length querysplits == 2)
    guard (length beforeQ == 7)
    guard (length afterQ == 3)
    guard (dm /= ".")
    return $ TRD ts rt dm
  where
    querysplits = T.split (=='?') line
    beforeQ     = splitOnSpace $ head querysplits
    afterQ      = splitOnSpace $ last querysplits
    rt          = last beforeQ
    ts          = head . splitOnDot $ head beforeQ
    dm          = head $ tail afterQ

splitOnSpace :: T.Text -> [T.Text]
splitOnSpace = T.split (== ' ')

splitOnDot :: T.Text -> [T.Text]
splitOnDot = T.split (== '.')

updateRTFS :: RecTypeFilterState -> TRD -> RecTypeFilterState
updateRTFS (RTFS _ (TRD _ _ dm)) trd = RTFS dm trd

eliminateTRD :: RecTypeFilterState -> Maybe TRD
eliminateTRD (RTFS dm' trd@(TRD _ _ dm)) =
    if dm == dm' then Nothing else Just trd

updateHitMap :: DomainProcState -> TRD -> DomainProcState
updateHitMap s trd = DPS tldM sldM tld sld dm
  where
    dm   = domainname trd
    dx   = splitOnDot dm
    tld  = Just . last $ init dx
    sld  = if length dx > 2 then Just . last . init $ init dx else Nothing
    tldM = incrementMap (tldMap s) tld
    sldM = incrementMap (sldMap s) sld

incrementMap :: DomainHitMap
             -> Maybe T.Text -- domain key if extracted
             -> DomainHitMap
incrementMap m Nothing  = m
incrementMap m (Just k) = HashMap.insertWith (+) k 1 m

outputLine :: DomainProcState -> Maybe T.Text
-- found SLD
outputLine (DPS _    sldM _          (Just sld) dm) = Just $
    prepareOutputLine sld (sldM HashMap.! sld) dm
-- at least found TLD
outputLine (DPS tldM _    (Just tld) Nothing    dm) = Just $
    prepareOutputLine tld (tldM HashMap.! tld) dm
-- no TLD (domainname == "."), the upstream should already yield an error
outputLine (DPS tldM _    Nothing    Nothing    dm) = Nothing

prepareOutputLine :: T.Text -- domain key: TLD or SLD
                  -> Integer -- hit
                  -> T.Text -- full domain name
                  -> T.Text -- output line
prepareOutputLine dk hit dm = T.concat [color co $ padding dk, "| ", dm]
  where
    co = colorHit hit
    pos = T.length dk - displayKeySpan
    padding x = T.justifyLeft displayKeySpan ' '
              $ if pos > 0 then T.concat [T.dropEnd (pos + 4) x, "..."] else x

color :: T.Text -- what color to use
      -> T.Text -- the input
      -> T.Text -- the output
color co t = T.concat [colorCode co, t, colorCode "default"]

printErr :: T.Text -> IO ()
printErr msg = TIO.hPutStrLn stderr $ T.concat
    [if isErrMsg msg then color "red" "[error] " else "[message] ", msg]

printErrInvalidRec :: T.Text -> IO ()
printErrInvalidRec line = printErr $ T.concat ["invalid record: ", line]


src/Config.hs
{-# LANGUAGE OverloadedStrings #-}

module Config
( monitoredRecType
, displayKeySpan
, colorHit
, colorCode
, isErrMsg
) where

import Data.Text (Text, concat)
import qualified Data.Text as T

monitoredRecType :: Text -> Bool
monitoredRecType rt = notElem rt ["PTR"]

displayKeySpan :: Int
displayKeySpan = 20

-- display color when hit more than
colorHit :: Integer -> Text
colorHit h
    | h == 1    = "yellow"
    | h < 16    = "default"
    | h < 128   = "grey"
    | otherwise = "black"

colorCode :: Text -> Text
colorCode "black"   = "\x1b[30m"
colorCode "red"     = "\x1b[31m"
colorCode "green"   = "\x1b[32m"
colorCode "yellow"  = "\x1b[33m"
colorCode "blue"    = "\x1b[34m"
colorCode "magenta" = "\x1b[35m"
colorCode "cyan"    = "\x1b[36m"
colorCode "white"   = "\x1b[37m"
colorCode "grey"    = "\x1b[90m"
colorCode _         = "\x1b[0m"

isErrMsg :: Text -> Bool
isErrMsg m = and [null $ T.breakOnAll x m | x <- nonTcpdumpErrMsgNeedles]

nonTcpdumpErrMsgNeedles :: [Text]
nonTcpdumpErrMsgNeedles = ["verbose output suppressed", "listening on"]
